File: Python/Problema 2/Problema_2.py
"""
Resolvendo o Problema 2
"""


from sys import exit

"""
Utilizando a função exit() para que não haja necessidade de rodar o resto do código a partir do momento em que 
se encontra que a tabela não é válida.
"""

board = [["5","3",".",".","7",".",".",".","."]
        ,["6",".",".","1","9","5",".",".","."]
        ,[".","9","8",".",".",".",".","6","."]
        ,["8",".",".",".","6",".",".",".","3"]
        ,["4",".",".","8",".","3",".",".","1"]
        ,["7",".",".",".","2",".",".",".","6"]
        ,[".","6",".",".",".",".","2","8","."]
        ,[".",".",".","4","1","9",".",".","5"]
        ,[".",".",".",".","8",".",".","7","9"]]

tabela_valida = True

# Verifica se há elementos repetidos em uma lista
def verificadorLinha(lista):
    for item in lista:
        if item == ".":
            continue
        else: 
            if lista.count(item) > 1:
                return False  
    return True

# Verifica todas as linhas do board
for lista in board: 
    if verificadorLinha(lista) == False:
        print(False)
        exit()


# Verifica todas as colunas do board
for i in range(9):
    coluna = []
    for j in range(9):
        coluna.append(board[j][i])
    if verificadorLinha(coluna) == False:
        print(False)
        exit()

# Verifica os quadrados do board
for i in [0,3,6]:
    for j in [0,3,6]:
        quadrado = [board[i][j], board[i+1][j], board[i+2][j], 
                 board[i][j+1], board[i+1][j+1], board[i+2][j+1], 
                 board[i][j+2], board[i+1][j+2], board[i+2][j+2]]
        if verificadorLinha(quadrado) == False:
            print(False)
            exit()

print(tabela_valida)
        

File: JavaScript/Problema 2/Problema_2.js
/*
Resolvendo o Problema 2
*/

const board = [["5", "3", ".", ".", "7", ".", ".", ".", "."]
             , ["6", ".", ".", "1", "9", "5", ".", ".", "."]
             , [".", "9", "8", ".", ".", ".", ".", "6", "."]
             , ["8", ".", ".", ".", "6", ".", ".", ".", "3"]
             , ["4", ".", ".", "8", ".", "3", ".", ".", "1"]
             , ["7", ".", ".", ".", "2", ".", ".", ".", "6"]
             , [".", "6", ".", ".", ".", ".", "2", "8", "."]
             , [".", ".", ".", "4", "1", "9", ".", ".", "5"]
             , [".", ".", ".", ".", "8", ".", ".", "7", "9"]]

let tabela_valida = true

// Verifica se há elementos repetidos em uma lista
function verificadorLinha(lista) {
    for (a = 0; a < lista.length; a++) {
        let contador = 0
        if (lista[a] === '.') {
            continue
        }
        else {
            for (b = 0; b < lista.length; b++) {
                if (lista[b] == lista[a]) { contador++ }
                if (contador > 1) {
                    return false

                }
            }
        }
    }
    return true
}

// Verifica todas as linhas do board
for (c = 0; c < board.length; c++) {
    if (verificadorLinha(board[c]) == false) {
        tabela_valida = false
    }
}

// Verifica todas as colunas do board
for (d = 0; d < 9; d++) {
    let coluna = []
    for (e = 0; e < 9; e++) {
        coluna.push(board[e][d])
        if (verificadorLinha(coluna) == false) {
            tabela_valida = false
        }
    }
}

// Verifica os quadrados do board
const ref_quad = [0, 3, 6]
for (i = 0; i < ref_quad.length; i++) {
    for (j = 0; j < ref_quad.length; j++) {
        let quadrado = [board[ref_quad[i]][ref_quad[j]], board[ref_quad[i] + 1][ref_quad[j]], board[ref_quad[i] + 2][ref_quad[j]],
        board[ref_quad[i]][ref_quad[j] + 1], board[ref_quad[i] + 1][ref_quad[j] + 1], board[ref_quad[i] + 2][ref_quad[j] + 1],
        board[ref_quad[i]][ref_quad[j] + 2], board[ref_quad[i] + 1][ref_quad[j] + 2], board[ref_quad[i] + 2][ref_quad[j] + 2]]
        if (verificadorLinha(quadrado) == false) {
            tabela_valida = false
        }
    }
}

console.log(tabela_valida)
